### ISSUE TITLE
[BUG] Why set `fan_in_fan_out` based on `len(peft_config.target_modules)` in `_prepare_lora_config`

### ISSUE THREAD

**Issue by silverriver** [huggingface/peft#193]
I am wondering why we should set `fan_in_fan_out` based on `len(peft_config.target_modules)` when we are using lora?

https://github.com/huggingface/peft/blob/64f63a7df2a02cfd144592d9aa9c871b59258c55/src/peft/mapping.py#L120

In my understanding, I can set any layer to lora layer, and control whether to use enable_lora or not

**Comment by pacman100** [comment]
it was a proxy to make users life easier who don't know when to set it properly, e.g., it is always used when q, k and v layers are combined and as such will be in cases when `len(peft_config.target_modules) == 1`.  `fan_in_fan_out` is used for example in GPT-2 with combined q,k and v. Will fix this in upcoming PR with relevant documentation.

**Comment by pacman100** [comment]
Yes, that makes sense, as mentioned above, this will be fixed.

**Comment by silverriver** [comment]
Moreover, `bloom`, `gpt_neox`, `gpt_neo` all use `nn.Linear` rather than `nn.Conv1D`

**Comment by silverriver** [comment]
> Yes, that makes sense, as mentioned above, this will be fixed.

Sounds great. When can I expect this PR be merged?

**Comment by silverriver** [comment]
Any update?

**Comment by pacman100** [comment]
This has been rectified in the main branch already, could you please try it
out and let us know

**Comment by Sanster** [comment]
> I have a self defined model that use a integrated query_key_value layer:
> 
> ```python
>         self.query_key_value = skip_init(
>             torch.nn.Linear,
>             hidden_size,
>             3 * self.inner_hidden_size,
>             bias=bias,
>             dtype=params_dtype,
>         )
> ```
> 
> I want to use LoRA to tune all Q,K,V matrics.
> 
> I use
> 
> ```python
>         peft_config = LoraConfig(
>             task_type=TaskType.CAUSAL_LM, merge_weights=False, r=8, lora_alpha=8, 
>             target_modules=['query_key_value'], bias='all', lora_dropout=0.1, fan_in_fan_out=False)
> ```
> 
> Based on the current implementation, I will get
> 
> ```python
> peft_config.enable_lora = [True, False, True]
> ```
> 
> in `_prepare_lora_config`
> 
> PEFT automaticlly fixes the K matrix. I think this is a BUG.

@silverriver Hi, I also encountered this problem. Have you observed whether fixing the k matrix will affect the accuracy of the model?
